Add `unsigned` support for numeric data types in MySQL

kamipo opened the pull request:
Example:

``` ruby
create_table :foos do |t|
  t.integer :unsigned_integer, unsigned: true
  t.bigint  :unsigned_bigint,  unsigned: true
  t.float   :unsigned_float,   unsigned: true
  t.decimal :unsigned_decimal, unsigned: true, precision: 10, scale: 2
end
```

In the case of using `unsigned` as the type:

``` ruby
create_table :foos do |t|
  t.unsigned_integer :unsigned_integer
  t.unsigned_bigint  :unsigned_bigint
  t.unsigned_float   :unsigned_float
  t.unsigned_decimal :unsigned_decimal, precision: 10, scale: 2
end
```

kamipo committed:
Add `unsigned` support for numeric data types in MySQL

Example:

    create_table :foos do |t|
      t.integer :unsigned_integer, unsigned: true
      t.bigint  :unsigned_bigint,  unsigned: true
      t.float   :unsigned_float,   unsigned: true
      t.decimal :unsigned_decimal, unsigned: true, precision: 10, scale: 2
    end

rafaelfranca commented:
@kamipo could you rebase it?

kamipo commented:
@rafaelfranca Rebased!

rafaelfranca commented:
Duh, I forgot to say that I agree with @sgrif here. I'd introduce new types instead of a new option.

kamipo committed:
Add `unsigned` types for numeric data types in MySQL

In the case of using `unsigned` as the type:

    create_table :foos do |t|
      t.unsigned_integer :unsigned_integer
      t.unsigned_bigint  :unsigned_bigint
      t.unsigned_float   :unsigned_float
      t.unsigned_decimal :unsigned_decimal, precision: 10, scale: 2
    end

rafaelfranca commented:
@kamipo does make sense to have `unsigned_float`, `unsigned_bigint` and `unsigned_decimal` types?

zzak commented:
I discussed with @rafaelfranca and we would prefer types over option, including primary key.
Similar to how types are handled in the postgres adapter.

/cc @sgrif

kamipo commented:
Even if the types is added, For primary key without option in MySQL, should need to handle `:auto_increment` option.

In the case of `:bigint` primary key in MySQL:

[active_record/connection_adapters/abstract_mysql_adapter.rb#L10](https://github.com/rails/rails/blob/ae5f2b4e79f3e41aad8280109d8bfc788a1a2733/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L10)

```ruby
      module ColumnMethods
        def primary_key(name, type = :primary_key, **options)
          options[:auto_increment] = true if type == :bigint
          super
        end
      end
```

In the case of `:uuid` primary key, set implicit default `'uuid_generate_v4()'` for without option.

[active_record/connection_adapters/postgresql/schema_definitions.rb#L14-L24](https://github.com/rails/rails/blob/ae5f2b4e79f3e41aad8280109d8bfc788a1a2733/activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb#L14-L24)

If we want to use other default instead, it is necessary to explicitly specify it.

```ruby
create_table :foos, id: :uuid, default: nil do |t|
end

# or

create_table :foos, id: :uuid, default: 'gen_random_uuid()' do |t|
end
```

Currently, the only way of `:bigint` primary key without `:auto_increment` in MySQL is the following:

```ruby
create_table :foos, id: :integer, limit: 8 do |t|
end
```

I think it is better to support the following way:

```ruby
create_table :foos, id: :bigint, default: nil do |t|
end
```

If we want to use unsigned primary key **with** `:auto_increment`, It considered the following ways:

```ruby
# unsigned integer primary key with auto_increment

## 1.
create_table :foos, unsigned: true do |t|
end

## 2.
create_table :foos, id: :unsigned_integer, auto_increment: true do |t|
end

# unsigned bigint primary key with auto_increment

## 3.
create_table :foos, id: :bigint, unsigned: true do |t|
end

## 4.
create_table :foos, id: :unsigned_bigint, auto_increment: true do |t|
end
```

I prefer 1. and 3. than 2. and 4.

If we want to use unsigned primary key **without** `:auto_increment`, It considered the following ways:

```ruby
# unsigned integer primary key without auto_increment

## 1.
create_table :foos, id: :integer, unsigned: true do |t|
end

## 2.
create_table :foos, id: :unsigned_integer do |t|
end

# unsigned bigint primary key without auto_increment

## 3.
create_table :foos, id: :integer, limit: 8, unsigned: true do |t|
end

## 4.
create_table :foos, id: :unsigned_bigint do |t|
end

## 5. (not support yet)
create_table :foos, id: :bigint, unsigned: true, default: nil do |t|
end
```

I think any good in this case.

kamipo commented:
r? @rafaelfranca 
I added `unsigned_float`, `unsigned_bigint` and `unsigned_decimal` types.
Could you review this changes?
